[3006.x] Fix 65777 : Default to utf-8 on all systems

twangboy opened the pull request:
### What does this PR do?
This attempts to fix some Unicode errors on Windows. Specifically when reading an INI file that contains Unicode characters. Though this should make Salt use utf-8 encoding by default. This attempts to be a more universal fix than just fixing a single module. We may need to address other modules that attempted to address Unicode issues with Python2 and others using MBCS on Windows.

### What issues does this PR fix or reference?
Fixes: #65777

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

dwoz commented on salt/__init__.py:
```diff
@@ -115,38 +115,29 @@ def __define_global_system_encoding_variable__():
+        # matches the filesystem encoding
         encoding = sys.stdin.encoding
 
     if not encoding:
```
Is there really ever a case where encoding would be Falsey here?

twangboy replied:
I can't think of one... no... I was just keeping the original code.

twangboy committed:
Rebase to 3006.x

twangboy committed:
Add encoding option to the ini module

Adds a new encoding parameter to the ini module to handle scenarios
where the ini file encoding is not utf-8.

twangboy committed:
Migrate ini_manage tests from unittests

twangboy committed:
Add some tests for unicode

Test reading and writing to utf-8, utf-16, utf-16-le and utf-32-le
ini files

twangboy committed:
Add documentation to the ini_manage module

Add missing documentation for all functions
Add documentation for new parameter `encoding`
Add Returns section to documentation

damon-atkins commented on salt/__init__.py:
```diff
@@ -131,60 +131,16 @@ def exec_module(self, module):
-            elif sys.platform.startswith("win"):
-                # Windows uses a configurable encoding; on Windows, Python uses the name “mbcs”
-                # to refer to whatever the currently configured encoding is.
-                encoding = "mbcs"
```
This section of code should remain for darwin and windows.

twangboy replied:
Starting with python 3.6 python is using `utf-8` internally and using the Unicode methods instead of the ANSI methods in Windows. As far as I can tell, mcbs only applies to Windows.

https://peps.python.org/pep-0529/

The original code was always hitting `locale.getdefaultlocale()[-1]` on Windows, which returns `cp1252`. Which was problematic because it doesn't not support Unicode and required that we pass encoding parameters everywhere. `utf-8` is a more sane default, especially since we are no longer supporting Python2.

The `sys.getdefaultencoding` function returns `utf-8` on darwin, so the behavior should be the same.

damon-atkins replied:
cp1252 is correct output for English windows.  Even if python is utf-8 Windows is not utf-8 for the most part.  And this is about https://github.com/saltstack/salt/issues/65777 which is one Windows application which can read utf-8.
The correct fix is to add to each state an output encoding.

I strongly advise that this change is tested on multiple installation of Windows with different default languages.  Or even reach out to everyone who has complain about a similar issue in Windows before.

Unfortunately we can not change all the 3 party application which expect a file in the encoding of mbcs  i.e. cp1252 and all other other encoding types for all the languages Windows supports.

damon-atkins replied:
> I also intend to revisit the to_mbcs functions in the registry module to see if they are still needed.
That was mostly about python 2 winapi functions only supporting mbcs,   where python 3 has a newer winapi which does not have the issue.

damon-atkins reviewed (changes requested):
When all the unicode work was completed some time ago.   I believe the aim was for all of the "Salt" source files to be read as uft-8 i.e. files in salt://   The output of these files on each platform is platform default encoding.
The default for https:// as source files would be what ever https indicates (defaulting to uft-8).

If you download a file first into the salt cache folder, do you read it as utf-8 or mbcs?

Is the assumption with this PR, is that every application written by 3rd parties, outside of salt control and outside of salt testing can read uft-8 files?  e.g. Non-English application.

I agree in every salt state you should be able to override the default **output** encoding.
Maybe their should be a `in-encoding` (source files e.g salt://) and `out-encoding` parameter.

However this code has work for a long long time.
```
        if not encoding:
            if sys.platform.startswith("darwin"):
                # Mac OS X uses UTF-8
                encoding = "utf-8"
            elif sys.platform.startswith("win"):
                # Windows uses a configurable encoding; on Windows, Python uses the name “mbcs”
                # to refer to whatever the currently configured encoding is.
                encoding = "mbcs"
            else:
                # On linux default to ascii as a last resort
                encoding = "ascii"
```
This was the first attempt https://github.com/saltstack/salt/pull/40307
